### PR TITLE
HW2

### PR DESCRIPTION
**Pull request by sarvesh2910**
This PR is about HW2.

### CODE REVIEW

File: hw2/01-coins.js [hunk 1]
```diff
@@ -1,7 +1,46 @@
...
+    answer = `${answer} ${dime} dime${dime >= 2 ? "s" : ""},`;
+  }
+  if (nickel) {
+    answer = `${answer} ${nickel} nickel${nickel >= 2 ? "s" : ""},`;
+  }
+  if (pennies) {
+    answer = `${answer} ${pennies} penn${pennies >= 2 ? "ies" : "y"}`;
+  }
+
+  console.log(answer);
```
Comment (prachi-kashyap) [inline hw2/01-coins.js]: 1. It seems you have not returned any value, which would end up showing undefined after each function call. Maybe you can update that.
2. If possible update the printing statement as suggested by Caterina(Not a major requirement)
Expected : $4.62 ==> 4 dollars, 2 quarters, 1 dime, 2 pennies
Actual : 4 dollars, 2 quarters, 1 dime, 2 pennies

Comment (sarvesh2910) [reply hw2/01-coins.js]: Not sure what happened here, but i've already made these changes.

Comment (sarvesh2910) [reply hw2/01-coins.js]: Got it. I committed them yesterday but forgot to push. lol
Thanks for suggestions though!

File: hw2/01-coins.js [hunk 2]
```diff
@@ -1,7 +1,42 @@
...
 const calculateChange = (input) => {
-  // Add your code here
+  let answer = `$${input} ==>`;
+  if (input <= 0) return `${answer} input should be positive`;
+  if (input > 10) return `${answer} Error: the number is too large`;
+
+  const coins = [25, 10, 5, 1];
+
+  // eslint-disable-next-line prefer-const
+  let [dollars, remaining] = input
```
Comment (ethanSE) [inline hw2/01-coins.js]: unfortunately this isn't working quite right. I'm not sure if splitting the string and parsing the two halves separately works as intended. $4.80 is being interpreted as 4 dollars and 8 cents.

test case: console.log(calculateChange(4.80))

File: hw2/01-coins.js [hunk 3]
```diff
@@ -1,7 +1,42 @@
...
+    answer = `${answer} ${dollars} dollar${dollars >= 2 ? "s" : ""},`;
+  }
+  if (quarters) {
+    answer = `${answer} ${quarters} quarter${quarters >= 2 ? "s" : ""},`;
+  }
+  if (dime) {
+    answer = `${answer} ${dime} dime${dime >= 2 ? "s" : ""},`;
+  }
+  if (nickel) {
+    answer = `${answer} ${nickel} nickel${nickel >= 2 ? "s" : ""},`;
```
Comment (ethanSE) [inline hw2/01-coins.js]: It probably doesn't matter but if there is some amount of nickles and no pennies then the output has a trailing comma

File: hw2/04-api.js [hunk 4]
```diff
@@ -1,7 +1,28 @@
...
 
 // Add your code here
-
+const getData = () => {
+  const list = document.getElementById("results");
+  fetch(url)
+    .then((data) => data.json())
+    .then((data) => {
+      list.innerHTML = "";
+      const allCountries = data;
```
Comment (ethanSE) [inline hw2/04-api.js]: You're not making any reference to data here other than creating a const reference called allCountries. ~~You could equivalently name your callback arg this and save some reading.~~
edit: sorry, the tone here sounded like a complaint. 

As the reader, when I'm reading this i'm mentally keeping track of the namespace. Often, the smaller this is the easier the task. When I see a new reference created to the callback arg I am then aware that there are two references to the same data and am keeping track of this mentally. I have to be aware that this data could still be accessed via the old reference. This is especially complicated when working with asynchronous code, or passing data through multiple functions.

```javascript
.then((allCountries) => {
      list.innerHTML = "";
      allCountries.forEach((item) => {
        const node = document.createElement("li");
        const text = `${item.name} - ${parseInt(
          item.population,
          10
        ).toLocaleString()}`;
        node.innerText = text;
        list.appendChild(node);
      });
    })
```

File: hw2/02-reverse.js [hunk 5]
```diff
@@ -1,3 +1,28 @@
...
+
+  if (input.value.length !== 8) {
+    textElement.classList.add("text-danger");
+    textElement.innerText = "Error : Please input an 8-digit number";
+    input.insertAdjacentElement("afterend", textElement);
+  } else {
+    const reversedInput = [...input.value].reverse().join("");
+    textElement.classList.add("text-success");
+    textElement.innerText = `${input.value} -> ${reversedInput}`;
+    input.insertAdjacentElement("afterend", textElement);
```
Comment (ethanSE) [inline hw2/02-reverse.js]: the behavior: 
```javascript
input.insertAdjacentElement("afterend", textElement);
```
is shared by both branches and could equivalently be placed after the if-else statement.

**Review by ethanSE — COMMENTED** [review]
Good overall. My only concern is the parsing of the coin input. I made a couple other comments that aren't that important

**Review by prachi-kashyap — APPROVED** [review]

File: hw2/01-coins.js (hunk 2 above)
Comment (sarvesh2910) [reply hw2/01-coins.js]: Thanks for pointing this out. This is weird JS bug, where JS removes the trailing zeroes. 
I could have solved it by multiplying `remaining ` by 10 if it less than 10. 
But I rewrote whole logic as i had to check for dynamic insertion of commas as well.

File: hw2/01-coins.js (hunk 3 above)
Comment (sarvesh2910) [reply hw2/01-coins.js]: Updated.

File: hw2/04-api.js (hunk 4 above)
Comment (sarvesh2910) [reply hw2/04-api.js]: Thanks. I totally agree with your thoughts. Somehow i missed it in code.

File: hw2/02-reverse.js (hunk 5 above)
Comment (sarvesh2910) [reply hw2/02-reverse.js]: Updated.

File: hw2/01-coins.js [hunk 6]
```diff
@@ -1,42 +1,42 @@
...
 
-  const coins = [25, 10, 5, 1];
+  let input100 = input * 100;
 
-  // eslint-disable-next-line prefer-const
-  let [dollars, remaining] = input
-    .toString()
-    .split(".")
-    .map((i) => parseInt(i, 10));
+  const coinChange = coinValue.map((coins) => {
```
Comment (ethanSE) [inline hw2/01-coins.js]: I'm nitpicking here but this variable "coins" is plural but the value represents the value of a single coin. Similarly, coinValue is a collection of coin values and has a singular name.

A pretty basic approach for naming these would be to use a plural and singular version of the same or similar name
coinValue -> coinValues
coins -> coinValue (or shorten to value?)

File: hw2/01-coins.js [hunk 7]
```diff
@@ -1,42 +1,42 @@
...
-  if (nickel) {
-    answer = `${answer} ${nickel} nickel${nickel >= 2 ? "s" : ""},`;
-  }
-  if (pennies) {
-    answer = `${answer} ${pennies} penn${pennies >= 2 ? "ies" : "y"}`;
-  }
-
-  return answer;
+  return `${answer} ${tempArray.join(", ")}`;
 };
```
Comment (ethanSE) [inline hw2/01-coins.js:41]: The variable named "answer" isn't very useful anymore and the name may no longer be accurate

**Review by ethanSE — APPROVED** [review]
Nice! This uses some more language features and is more DRY.

File: hw2/01-coins.js (hunk 7 above)
Comment (sarvesh2910) [reply hw2/01-coins.js:41]: updated.

File: hw2/01-coins.js (hunk 6 above)
Comment (sarvesh2910) [reply hw2/01-coins.js]: I agree with you. Again i missed this simple things. lol. 

Updated.

**Pull request merged**